File: python/IceStorm/clock/Subscriber.py
#!/usr/bin/env python
# **********************************************************************
#
# Copyright (c) 2003-2018 ZeroC, Inc. All rights reserved.
#
# **********************************************************************

import signal, sys, Ice, IceStorm, getopt

Ice.loadSlice('Clock.ice')
import Demo

class ClockI(Demo.Clock):
    def tick(self, date, current):
        print(date)

def usage():
    print("Usage: " + sys.argv[0] +
          " [--batch] [--datagram|--twoway|--ordered|--oneway] [--retryCount count] [--id id] [topic]")

def run(communicator, args):
    try:
        opts, args = getopt.getopt(args[1:], '', ['datagram', 'twoway', 'oneway', 'ordered', 'batch',
                                                  'retryCount=', 'id='])
    except getopt.GetoptError:
        usage()
        return 1

    batch = False
    option = "None"
    topicName = "time"
    id = ""
    retryCount = ""

    for o, a in opts:
        oldoption = option
        if o == "--datagram":
            option = "Datagram"
        elif o == "--twoway":
            option = "Twoway"
        elif o == "--ordered":
            option = "Ordered"
        elif o == "--oneway":
            option = "Oneway"
        elif o == "--batch":
            batch = True
        elif o == "--id":
            id = a
        elif o == "--retryCount":
            retryCount = a
        if oldoption != option and oldoption != "None":
            usage()
            return 1

    if len(args) > 1:
        usage()
        return 1

    if len(args) > 0:
        topicName = args[0]

    if len(retryCount) > 0:
        if option == "None":
            option = "Twoway"
        elif option != "Twoway" and option != "Ordered":
            print(sys.argv[0] + ": retryCount requires a twoway proxy")
            return 1

    if batch and (option in ("Twoway", "Ordered")):
        print(sys.argv[0] + ": batch can only be set with oneway or datagram")
        return 1

    manager = IceStorm.TopicManagerPrx.checkedCast(communicator.propertyToProxy('TopicManager.Proxy'))
    if not manager:
        print(args[0] + ": invalid proxy")
        return 1

    #
    # Retrieve the topic.
    #
    try:
        topic = manager.retrieve(topicName)
    except IceStorm.NoSuchTopic as e:
        try:
            topic = manager.create(topicName)
        except IceStorm.TopicExists as ex:
            print(sys.argv[0] + ": temporary error. try again")
            return 1

    adapter = communicator.createObjectAdapter("Clock.Subscriber")

    #
    # Add a servant for the Ice object. If --id is used the identity
    # comes from the command line, otherwise a UUID is used.
    #
    # id is not directly altered since it is used below to detect
    # whether subscribeAndGetPublisher can raise AlreadySubscribed.
    #

    subId = Ice.Identity()
    subId.name = id
    if len(subId.name) == 0:
        subId.name = Ice.generateUUID()
    subscriber = adapter.add(ClockI(), subId)

    #
    # Activate the object adapter before subscribing.
    #
    adapter.activate()

    qos = {}
    if len(retryCount) > 0:
        qos["retryCount"] = retryCount

    #
    # Set up the proxy.
    #
    if option == "Datagram":
        if batch:
            subscriber = subscriber.ice_batchDatagram()
        else:
            subscriber = subscriber.ice_datagram()
    elif option == "Twoway":
        # Do nothing to the subscriber proxy. Its already twoway.
        pass
    elif option == "Ordered":
        # Do nothing to the subscriber proxy. Its already twoway.
        qos["reliability"] = "ordered"
    elif option == "Oneway" or option == "None":
        if batch:
            subscriber = subscriber.ice_batchOneway()
        else:
            subscriber = subscriber.ice_oneway()

    try:
        topic.subscribeAndGetPublisher(qos, subscriber)
    except IceStorm.AlreadySubscribed as ex:
        # If we're manually setting the subscriber id ignore.
        if len(id) == 0:
            raise
        print("reactivating persistent subscriber")

    signal.signal(signal.SIGINT, lambda signum, frame: communicator.shutdown())
    communicator.waitForShutdown()

    #
    # Unsubscribe all subscribed objects.
    #
    topic.unsubscribe(subscriber)

    return 0

status = 0

#
# Ice.initialize returns an initialized Ice communicator,
# the communicator is destroyed once it goes out of scope.
#
with Ice.initialize(sys.argv, "config.sub") as communicator:
    status = run(communicator, sys.argv)
sys.exit(status)

File: cpp/IceStorm/clock/Subscriber.cpp
// Copyright (c) ZeroC, Inc.

#include "Clock.h"

#include <Ice/Ice.h>
#include <IceStorm/IceStorm.h>

#include <iostream>

using namespace std;
using namespace Demo;

/// ClockI is an Ice servant that implements Slice interface Clock.
class ClockI final : public Clock
{
public:
    void tick(string time, const Ice::Current&) final { cout << time << endl; }
};

void
usage(const char* name)
{
    cerr << "Usage: " << name
         << " [--batch] [--datagram|--twoway|--ordered|--oneway] [--retryCount count] [--id id] [topic]" << endl;
}

int
main(int argc, char* argv[])
{
#ifdef ICE_STATIC_LIBS
    Ice::registerIceUDP();
#endif

    // CtrlCHandler is a helper class that handles Ctrl+C and similar signals. It must be constructed at the beginning
    // of the program, before creating an Ice communicator or starting any thread.
    Ice::CtrlCHandler ctrlCHandler;

    // Create an Ice communicator to initialize the Ice runtime.
    const Ice::CommunicatorHolder communicatorHolder{argc, argv, "config.sub"};
    const Ice::CommunicatorPtr& communicator = communicatorHolder.communicator();

    // Parse command-line options.

    auto args = Ice::argsToStringSeq(argc, argv);
    args = communicator->getProperties()->parseCommandLineOptions("Clock", args);
    Ice::stringSeqToArgs(args, argc, argv);

    bool batch = false;
    enum class Option : uint8_t
    {
        None,
        Datagram,
        Twoway,
        Oneway,
        Ordered
    };

    Option option = Option::None;
    string topicName = "time";
    string id;
    string retryCount;
    int i;

    for (i = 1; i < argc; ++i)
    {
        const string optionString = argv[i];
        const Option oldoption = option;
        if (optionString == "--datagram")
        {
            option = Option::Datagram;
        }
        else if (optionString == "--twoway")
        {
            option = Option::Twoway;
        }
        else if (optionString == "--oneway")
        {
            option = Option::Oneway;
        }
        else if (optionString == "--ordered")
        {
            option = Option::Ordered;
        }
        else if (optionString == "--batch")
        {
            batch = true;
        }
        else if (optionString == "--id")
        {
            ++i;
            if (i >= argc)
            {
                usage(argv[0]);
                return 1;
            }
            id = argv[i];
        }
        else if (optionString == "--retryCount")
        {
            ++i;
            if (i >= argc)
            {
                usage(argv[0]);
                return 1;
            }
            retryCount = argv[i];
        }
        else if (optionString.substr(0, 2) == "--")
        {
            usage(argv[0]);
            return 1;
        }
        else
        {
            topicName = argv[i++];
            break;
        }

        if (oldoption != option && oldoption != Option::None)
        {
            usage(argv[0]);
            return 1;
        }
    }

    if (i != argc)
    {
        usage(argv[0]);
        return 1;
    }

    if (batch && (option == Option::Twoway || option == Option::Ordered))
    {
        cerr << argv[0] << ": batch can only be set with oneway or datagram" << endl;
        return 1;
    }

    // Create the topic manager proxy.
    auto manager = communicator->propertyToProxy<IceStorm::TopicManagerPrx>("TopicManager.Proxy");
    if (!manager)
    {
        cerr << argv[0] << ": invalid proxy" << endl;
        return 1;
    }

    // Retrieve the topic from IceStorm.
    optional<IceStorm::TopicPrx> topic;
    try
    {
        topic = manager->retrieve(topicName);
    }
    catch (const IceStorm::NoSuchTopic&)
    {
        try
        {
            topic = manager->create(topicName);
        }
        catch (const IceStorm::TopicExists&)
        {
            cerr << argv[0] << ": temporary failure. try again." << endl;
            return 1;
        }
    }

    // Create the object adapter that hosts the subscriber servant.
    auto adapter = communicator->createObjectAdapter("Clock.Subscriber");

    // Add a servant to the object adapter. If --id is used the identity comes from the command line, otherwise a UUID
    // is used. id is not directly altered since it is used below to detect whether subscribeAndGetPublisher can throw
    // AlreadySubscribed.
    Ice::Identity subId{.name = id};
    if (subId.name.empty())
    {
        subId.name = Ice::generateUUID();
    }
    Ice::ObjectPrx subscriber = adapter->add(make_shared<ClockI>(), subId);

    // Activate the object adapter before subscribing.
    adapter->activate();

    IceStorm::QoS qos;
    if (!retryCount.empty())
    {
        qos["retryCount"] = retryCount;
    }

    // Set up the subscriber proxy.
    if (option == Option::Datagram)
    {
        if (batch)
        {
            subscriber = subscriber->ice_batchDatagram();
        }
        else
        {
            subscriber = subscriber->ice_datagram();
        }
    }
    else if (option == Option::Twoway)
    {
        // Do nothing to the subscriber proxy. Its already twoway.
    }
    else if (option == Option::Ordered)
    {
        // Do nothing to the subscriber proxy. Its already twoway.
        qos["reliability"] = "ordered";
    }
    else if (option == Option::Oneway || option == Option::None)
    {
        if (batch)
        {
            subscriber = subscriber->ice_batchOneway();
        }
        else
        {
            subscriber = subscriber->ice_oneway();
        }
    }

    try
    {
        // Register the subscriber with the IceStorm topic.
        topic->subscribeAndGetPublisher(qos, subscriber);
    }
    catch (const IceStorm::AlreadySubscribed&)
    {
        // This should never occur when subscribing with a UUID.
        assert(!id.empty());
        cout << "reactivating persistent subscriber" << endl;
    }

    // Wait until the user presses Ctrl+C.
    int signal = ctrlCHandler.wait();
    cout << "Caught signal " << signal << ", exiting..." << endl;

    // Initiate shutdown and wait for all dispatches to complete.
    communicator->shutdown();
    communicator->waitForShutdown();

    // Unsubscribe from the topic in IceStorm.
    topic->unsubscribe(subscriber);
    return 0;
}
